service/history/workflowResetor.go

```diff
@@ -431,15 +455,78 @@ func (w *workflowResetorImpl) generateReplicationTasksForReset(terminateCurr boo
 	return repTasks
 }
 
-func (w *workflowResetorImpl) replayReceivedSignals(receivedSignals []*workflow.HistoryEvent, msBuilder mutableState) {
+// replay signals in the base run, and also signals in all the runs along the chain of contineAsNew
+func (w *workflowResetorImpl) replayReceivedSignals(ctx context.Context, receivedSignals []*workflow.HistoryEvent, continueRunID string, newMutableState, currMutableState mutableState) error {
 	for _, se := range receivedSignals {
 		sigReq := &workflow.SignalWorkflowExecutionRequest{
 			SignalName: se.GetWorkflowExecutionSignaledEventAttributes().SignalName,
 			Identity:   se.GetWorkflowExecutionSignaledEventAttributes().Identity,
 			Input:      se.GetWorkflowExecutionSignaledEventAttributes().Input,
 		}
-		msBuilder.AddWorkflowExecutionSignaled(sigReq.GetSignalName(), sigReq.GetInput(), sigReq.GetIdentity())
+		newMutableState.AddWorkflowExecutionSignaled(sigReq.GetSignalName(), sigReq.GetInput(), sigReq.GetIdentity())
+	}
+	for {
```

Comment:
the continue as new has a ton of ambiguity, like 
1. parent workflow, after continue as new is no longer the parent of child workflows created in the prev run
2. child workflow, after continue as new is still the child of the workflow which created it.

garbage collection on signals which does not belonging to the base workflow is wrong, since we let customer specify the workflow id and run id..

Reply:
Issue created as our offline discussion with @samarabbas  : https://github.com/uber/cadence/issues/1412